torchx/schedulers/ray/ray_driver.py

```diff
@@ -77,77 +111,184 @@ def load_actor_json(filename: str) -> List[RayActor]:
         return actors
 
 
-def create_placement_group(replicas: List[RayActor]) -> PlacementGroup:
+def create_placement_group_async(replicas: List[RayActor]) -> PlacementGroup:
+    """return a placement group reference, the corresponding placement group could be scheduled or pending"""
     bundles = []
     for replica in replicas:
         bundles.append({"CPU": replica.num_cpus, "GPU": replica.num_gpus})
 
-    # To change the strategy type
-    # refer to available options here https://docs.ray.io/en/latest/placement-group.html#pgroup-strategy
     pg = ray.util.placement_group(bundles, strategy="SPREAD")
+    return pg
 
-    _logger.info("Waiting for placement group to start.")
-    ready = pg.wait(timeout_seconds=100)
 
-    if not ready:  # pragma: no cover
-        raise TimeoutError(
-            "Placement group creation timed out. Make sure "
-            "your cluster either has enough resources or use "
-            "an autoscaling cluster. Current resources "
-            "available: {}, resources requested by the "
-            "placement group: {}".format(ray.available_resources(), pg.bundle_specs)
+@dataclass
+class ActorInfo:
+    pg: PlacementGroup
+    replica: RayActor
+    actor: CommandActor
+
+
+class RayDriver:
+    def __init__(self, replicas: List[RayActor]) -> None:
+        self.replicas = replicas
+        self.master_node_id: Optional[str] = None
+        self.rank_0_address: Optional[str] = None
+        self.rank_0_port: Optional[int] = None
+        self.min_nnodes, self.max_nnodes = parse_nnodes_rep(replicas)  # pyre-ignore
+
+        self.placement_groups: List[PlacementGroup] = []
+        self.actor_info_of_id: Dict[str, ActorInfo] = {}
+        self.active_tasks: List["ray.ObjectRef"] = []
+
+    def init_placement_groups(self) -> None:
+        """Initialize all placement groups needed for this job"""
+        replica_ix_of_pg: List[int] = [0] + list(
```

Comment:
it might be better to explicitly make the calls here. + we need to wait for the first placement group before creating the remainders otherwise the smaller groups might get scheduled first (unless ray does fifo queuing, which if it does we should document that)

```
initial = create_placement_group_async(self.replicas[0: self.min_nodes])
pg.wait(itimeout_seconds=...)
self.groups = []
for range(self.min_nodes, self.max_nodes):
    self.groups.append(create_placement_group_async(self.replicas[i:i+1]))
```

Reply:
Ray doesn't do FIFO, will change.